### ISSUE TITLE
Speed up Azure Pipelines when doing a release

### ISSUE THREAD

**Issue by bmw** [certbot/certbot#8139]
When we create a release, we trigger a CI build for each of these events:

* A build for the tag
* A build for the a.b.x branch
* A build for the PR to update certbot-auto

In Travis, all jobs in a build would run to completion before any jobs in a subsequent build are started.

This is not the case in Azure Pipelines. There we make heavy use of stages allowing some parts of the build to wait on others to complete. For example, we wait on snaps to be built to test them. It appears that jobs that are initially blocked on another job have a lower queue priority than jobs triggered by a completely different build. As a result, the sequence of tests in runs in Azure is:

1. The majority of the tests on the tag run.
2. The majority of the test on the a.b.x branch run.
3. The tests on the PR run.
4. The initially blocked builds from the tag ruin.
5. The initially blocked builds from the PR run.

As a result of this, waiting for the builds on the tag to finish so you can finish the release process takes much longer. The Travis builds we had in the past for a release took ~30 mins. The Azure release took over more than 2.5 times as long. See https://dev.azure.com/certbot/certbot/_build/results?buildId=2283&view=results.

The best ideas I currently have to solve this doing one of:

1. Don't create the `*.x` branch until it is necessary to do a point release. If we go this route, I think we should document how to create this branch probably with an automated script to simplify the process.
2. (Doesn't work ☹️ )Make the tests for the tag and the `*.x` run in the same Azure Pipeline and set `batch=True`. See https://docs.microsoft.com/en-us/azure/devops/pipelines/repos/github?view=azure-devops&tabs=yaml#batching-ci-runs.

EDIT: I largely rewrote this to provide more context.

**Comment by adferrand** [comment]
Aside the solution to find here, it would like to note that the overal time of this specific build is not normal. As one can see on this run for instance https://dev.azure.com/certbot/certbot/_build/results?buildId=2285 that the `TestAndPackage` step takes around 35 min usually, while on the run https://dev.azure.com/certbot/certbot/_build/results?buildId=2283 it took 1h10.

**Comment by bmw** [comment]
Sure, although I would like to note that I think we should expect it to be a normal occurrence every time we do a release because of all of the builds that get triggered that I described in my initial post.

**Comment by adferrand** [comment]
Of course, and we can expect more if we add the dns plugin snaps.

For the propositions, I like 2 because it would smartly use the fact that tags and branches can actually refer to the same commit.

**Comment by bmw** [comment]
> For the propositions, I like 2 because it would smartly use the fact that tags and branches can actually refer to the same commit.

Would it not even run a build for the `1.6.x` branch then if the tag started first?

For what it's worth, before writing this issue, I started playing with this approach and even went so far to combine all of our slight variations of the full test suite into one pipeline. An untested and quickly written version of that is:
```
# Advanced pipeline running our full test suite and deploying artifacts when appropriate.
trigger:
  branches:
    include:
      - '*.x'
      # When changing these triggers, please ensure the documentation under
      # "Running tests in CI" is still correct.
      - test-*
  tags:
    include:
      - v*
pr: none
schedules:
  - cron: "30 4 * * *"
    displayName: Nightly build
    branches:
      include:
      - master
    always: true

stages:
  - template: templates/stages/test-and-package-stage.yml
  - template: templates/stages/changelog-stage.yml
  - ${{ if not(or(startsWith(variables['Build.SourceBranchName'], 'test-'), endsWith(variables['Build.SourceBranchName'], '.x'))) }}:
    - template: templates/stages/deploy-stage.yml
  # Don't send Mattermost messages about failures on test- branches.
  - ${{ if not(startsWith(variables['Build.SourceBranchName'], 'test-')) }}:
    - template: templates/stages/notify-failure-stage.yml

```

**Comment by bmw** [comment]
Unfortunately, the 2nd idea doesn't work. Maybe `batch` works if each build has the same trigger such as multiple commits to the same branch, but it doesn't work across triggers (which kind of makes sense given its purpose).

For now, I'm tempted to write a PR to disable builds on commits to `*.x` branches and update the release instructions to trigger these builds manually before point releases.

Any objections or other suggestions @ohemorange?

**Comment by ohemorange** [comment]
I'm good with that suggestion. If we do that, can we have nightly builds on `*.x` branches? I think it's pretty common to come back the next day to find we need to do a point release, and we might want to have run tests the night before.

**Comment by bmw** [comment]
That's an interesting idea!

In the case we have to come in and do a point release, I think we'd first need to write and commit the bug fixes for the point release and then we'd want to rerun all the tests with those commits, but it'd certainly be nice to know that the point release branch is always in a working state. It's not uncommon for the environment around Certbot to change causing our tests to break which we then fix on `master`, but we'd have to also make these changes to the point release branch in order to run tests for a release. Ensuring this is always taken care of in case we do have to do a point release gives us one less potential hurdle to jump over when we are trying to put out a point release which is nice.

I don't think it's entirely trivial to do this since I think we'd have to update the branch to run tests nightly on as part of the release process since we have at least 38 branches matching the pattern `*.x` right now and we'd want to avoid conflicts with things like test farm tests in our normal nightly branch but it could be done.

(As a more radical idea that I haven't fully thought through, what if we just stopped doing point releases and always released off of `master`? We'd _maybe_ have to be more careful about adding code to `master` we're not ready to release than we are now (although I think we're already pretty conservative about this), but in many ways I think it may be easier for us.)

I'm happy to think/talk about this more, but I started thinking about this issue again to avoid headaches for you with the release next week and I don't think we're losing anything in the area you brought up by no longer building on pushes to `*.x`. The CI build on the tag is a superset of the tests on pushes to `*.x` and we don't run nightly tests on point releases now. [Our PR test suite is still run on PRs for `*.x` branches](https://github.com/certbot/certbot/blob/df1ca726f975b4f2cf96ada29785a4d35027e5f9/.azure-pipelines/main.yml#L4). I think all we're losing is that we have to manually trigger tests before a point release which you seem OK with. With this in mind, I created https://github.com/certbot/certbot/pull/8185.

**Comment by ohemorange** [comment]
Hm, if it's not easy to do the nightly, I'm not sure it's worth it, especially since we would want to rerun tests anyway.

> (As a more radical idea that I haven't fully thought through, what if we just stopped doing point releases and always released off of master? We'd maybe have to be more careful about adding code to master we're not ready to release than we are now (although I think we're already pretty conservative about this), but in many ways I think it may be easier for us.)

Ehhh I'd be very worried about this. More commits in between means worrying about new potential changes that could also cause problems; if we're doing the point release, the last thing we want is to potentially introduce something new that will necessitate *another* point release.

**Comment by bmw** [comment]
A new idea I had here is we could create a `stable` branch instead of separate `a.b.x` branches. As part of the release, we'd create a PR against this branch instead of creating/modifying a `a.b.x` branch. This would allow us to fairly easily run nightly tests on the branch and we could probably even add back runs of the full test suite on pushes to this branch since the PR probably usually wouldn't land until after the release build is finished. We could have the release instructions include a reminder about this and/or move the step about creating this PR to the end of the release instructions to help ensure this.

The only downside I see is I think there will occasionally be merge conflicts in the case of PRs like https://github.com/certbot/certbot/pull/7523 which were made to fix tests on the `0.40.x` branch and hadn't been merged back into master yet, but I think absent changes like this, new merge conflicts actually shouldn't happen.

What do you think? If you think that's a good idea, I'll create a new issue about this, probably not in any milestone but with the `significant` label, although I'm open to other suggestions. If you don't like this idea for some reason, I personally think we should just forget about this problem for now.

> Ehhh I'd be very worried about this. More commits in between means worrying about new potential changes that could also cause problems; if we're doing the point release, the last thing we want is to potentially introduce something new that will necessitate another point release.

Fair enough! Those are certainly good concerns.

**Comment by ohemorange** [comment]
Wait a second -- how would that be different from making a new `a.b.x` branch? Since we don't create the `a.b.x` branch until after the release is done usually, would we even be running into this problem regularly?

**Comment by bmw** [comment]
It's different because the branch name is consistent allowing us to easily configure nightly tests.

I think we would hit the problem regularly in our previous setup because I think the time between these two events is shorter than the half hour it takes the release build to run:

1. Pushing the tag to GitHub which triggers the build we need to wait on.
2. Pushing the `a.b.x` branch.

Pushing the `a.b.x` branch before https://github.com/certbot/certbot/pull/8185 landed caused our full test suite to be queued which the later parts of the release build would then be blocked on.

It's true we could solve this latter problem and original purpose of this issue though by delaying creating that branch until the build for the tag is finished. Personally though, I think I'd rather just create the PR against `stable` while I'm dealing with tags, branches, and `master` PRs and then take the 10 minute delay caused by our PR tests running, but if you preferred, you could always delay creating the PR until the release build finished to get the same behavior.

**Comment by ohemorange** [comment]
Will opening a PR against the existing `stable` branch not trigger the same tests?

**Comment by bmw** [comment]
No. [The tests we run against PRs](https://github.com/certbot/certbot/blob/4d3f6c23be7a28c1cd7f4498ee7413b68e218af1/.azure-pipelines/main.yml) is a much smaller set of tests than our full test suite.

**Comment by ohemorange** [comment]
And when you create a new branch it runs the full test suite, not the PR tests?

**Comment by ohemorange** [comment]
Could we change that?

**Comment by bmw** [comment]
We could, but what's the value of running our PR tests on pushes to the `*.x` branch? I think we'd want to run the full test suite before releasing any changes made to those branches.

**Comment by ohemorange** [comment]
Basically I feel like we could solve the problem if we turn off tests for the initial commit, no?

**Comment by bmw** [comment]
If there was a way to only turn off tests for the initial commit and not subsequent commits, it'd solve this issue. I'm not aware of a sane way to do that though.

https://github.com/certbot/certbot/pull/8185 turned off tests for all commits at the cost of having to trigger the tests for point releases manually.

**Comment by ohemorange** [comment]
Ah, unfortunate. Hey, what if we create the branch at the *very beginning* of the release process? By the time offline signing and test farm tests and everything are done, those tests will probably be done. And then at the end, we can submit a PR against the new branch with the release commits.

**Comment by bmw** [comment]
That'd work with the only downsides I see being:

1. We still can't have nightly tests on point release branches.
2. That branch is created before we know the release was actually successful, however, I think this is fine as long as we make it clear it's OK in the release instructions. When redoing the release, I think we'd just skip this step and the PR at the end would contain extra commits.

What's your concern about the `stable` branch idea? Merge conflicts/the branches diverging? There very well may be something I'm overlooking, but I think that should be a pretty minor problem. I think we deal with similar merge conflicts now when merging point releases into master and the only new conflicts I'm aware of are if we backport a test fix to the `stable` branch and don't do a point release.

Keeping in mind that the only burden added by https://github.com/certbot/certbot/pull/8185 was we have to manually create a `test-` branch with the commit we want to test before point releases, I personally think we should either:

1. Figure out a way to run nightly tests on point release branches.
2. Close this issue and move on for now.

I personally think the burden of creating a `test-` branch is as easy or easier than adding extra push steps/PRs to the release process. In the interest of simplicity, I'm leaning towards the latter.

**Comment by ohemorange** [comment]
> What's your concern about the stable branch idea?

1. Changes made during the last release will cause potential merge conflicts during this one, which seems unfair. The thought of having another thing to deal with at release time that isn't guaranteed to go smoothly gives me a knee-jerk reaction away from it.
2. No convenient history. For just looking at code it wouldn't be so bad, but it makes it more annoying if we ever want to do a point release on an older branch, for, say, distros or whatever.

>  I think we deal with similar merge conflicts now when merging point releases into master

Yes, but those happen *after* the main part of the release is done, whereas this would be a blocker for doing other things.

That being said, I don't think it's that bad, I was just hoping to find an even cleaner solution.

Yeah, I agree, we should just close this. I was hoping we might find something very clean and simple that puts us basically where we were before in terms of features, but given that we haven't really, the status quo isn't bad.

**Comment by bmw** [comment]
Closing as fixed by https://github.com/certbot/certbot/pull/8185.
